Simplified Operation Continuation : Basic Infrastructure

pveentjer opened the pull request:
Goal of this PR is to introduce the 'operation continuation' with the minimum number of changed files.

Introducing Operation continuation requires returning some information about the state of the execution e.g. if blocking is needed or the operation is offloaded. This would require a massive number of changes since the signature of the Operation.run needs to be modified. 

This PR introduces a new Operation.call method that can be overridden en can return this state. However, if this method isn't implemented, it forwards the call to the original Operation.run. This means that all the old code will keep working without modification and the new more advanced call method can be used when needed. 

The old methods will be deprecated so that in the future we'll get rid of this duplicate behavior. But for the time being makes integration of the operation continuation with the minimum number of changes possible.

pveentjer commented:
verify

devOpsHazelcast commented:
Test FAILed.

pveentjer commented:
verify

devOpsHazelcast commented:
Test PASSed.

pveentjer commented on hazelcast/src/main/java/com/hazelcast/executor/impl/operations/AbstractCallableTaskOperation.java:
```diff
@@ -58,32 +50,25 @@ public AbstractCallableTaskOperation(String name, String uuid, Data callableData
-     * @return
-     */
-    private Callable getCallable() {
-        try {
```
If an exception is thrown while deserializing, it will be caught by the OperationRunner and no matter the 'hasResponse' result, a response is going to be send. This is changed in 3.9, so we don't need to trap the exception in this operation to get is shipped.

pveentjer commented on hazelcast/src/main/java/com/hazelcast/spi/Operation.java:
```diff
@@ -97,7 +100,21 @@ public void beforeRun() throws Exception {
+    public void run() throws Exception {
+    }
+
+    public CallStatus call() throws Exception {
```
The magic parts where the old Operation.run keeps on working, but the new call functionality can be used as well.

tombujok commented on hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/OperationRunnerImpl.java:
```diff
@@ -203,6 +198,43 @@ public void run(Operation op) {
+                break;
+            case OFFLOADED:
+                op.afterRun();
+                break;
```
Do we want to run the afterRun here? not after the result is returned?

pveentjer replied:
Good point: Currently we are free to choose.

It makes sense to execute it since the run/call was executed, so I would assume the afterRun would then be executed as well.

However.. in case of a true continuation, the run method could be called many times, so the afterRun could be called many times. If e,g, cleanup or statistics updating would be done in the afterRun, this would be executed more than once which can lead to all kinds of problems.

So afterRun has become ambiguous.

tombujok commented on hazelcast/src/main/java/com/hazelcast/map/impl/operation/EntryOperation.java:
```diff
@@ -367,29 +377,6 @@ private void unlockOnly(final Object result, String caller, long threadId, long
 
-    @Override
-    public void onExecutionFailure(Throwable e) {
-        if (offloading) {
```
This was always the tricky part, if returnsResponse == false and this means that the caller won't be notified in case of exceptions, that may be retryable. How is this handled in the new code? Don't quite see that anywhere.

pveentjer replied:
There is a change in 3.9 that any exception during operation execution is always send no matter what. 

https://github.com/hazelcast/hazelcast/pull/11282

pveentjer commented:
Error handling of exceptions on the Operation.run/call doesn't require any additional modification. Since 3.9 always a response is send to the caller.

```
private void handleOperationError(Operation operation, Throwable e) {
        if (e instanceof OutOfMemoryError) {
            OutOfMemoryErrorDispatcher.onOutOfMemory((OutOfMemoryError) e);
        }
        try {
            operation.onExecutionFailure(e);
        } catch (Throwable t) {
            logger.warning("While calling 'operation.onFailure(e)'... op: " + operation + ", error: " + e, t);
        }

        operation.logError(e);

        if (operation instanceof Backup) {
            failedBackupsCounter.inc();
            return;
        }

        // a response is send regardless of the Operation.returnsResponse method because some operation do want to send
        // back a response, but they didn't want to send it yet but they ran into some kind of error. If on the receiving
        // side no invocation is waiting, the response is ignored.
        sendResponseAfterOperationError(operation, e);
    }
```

I don't have a strong opinion yet about automatic registration of asynchronous tasks. It is easy to automatically register, but there is currently no automatic hook to unregister. I would like to keep the register/unregister symetrics. So either both go automatic, or none go automatic.

I also don't have a strong opinion yet about response sending. If an offloaded task needs to send a response, it can just use the response handler of the original method.

tombujok commented on hazelcast/src/main/java/com/hazelcast/spi/CallStatus.java:
```diff
@@ -0,0 +1,58 @@
+ *
+ */
+public enum CallStatus {
+
```
When I read the enum it bothers me a bit that some statuses indicate that the operations has finished implicitly and some not. It makes it ambiguous, e.g. "NO_RESPONSE", can be read that there is no response yet. 
Could we try to formulate it in a clearer way. Just giving it a shot :

- COMPLETED_HAS_RESPONSE
- COMPLETED_NO_RESPONSE
- OFFLOADED
- SHOULD_WAIT

tombujok commented on hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/OperationRunnerImpl.java:
```diff
@@ -203,6 +198,43 @@ public void run(Operation op) {
+                afterRun(op);
+                break;
+            case WAIT:
+                nodeEngine.getOperationParker().park((BlockingOperation) op);
```
Could we extract it to a method - would be so nice just to see `park(op)`;

tombujok replied:
This one is just a minor thing. Would just look cleaner as a method call, but I'm accepting the PR anyway.

tombujok commented:
Looks good overall. Just minor comments about the naming, etc.

pveentjer commented:
I think the ambiguous nature of the afterRun can be fixed by leaving it as it is. And when the Offloaded is introduced, we can introduce a 'onStart' and 'onCompletion' on the offloaded itself. 

So afterRun will be executed every time after a run, the same as before run will be executed before any run.

You concerns regarding registration of async operations and responses, can be taken care of by the new Offloaded class (I'll make this phase 2).

tombujok commented on hazelcast/src/main/java/com/hazelcast/spi/CallStatus.java:
```diff
@@ -0,0 +1,66 @@
+ * release the operation thread so that other operations can be interleaved.
+ */
+public enum CallStatus {
+
```
looks better now 👍

jerrinot commented:
it looks good. the `@Depracated` methods should provide an explanation why they are deprecated and provide an explanation what's the right way of doing things.

pveentjer commented:
Since yet not everything is in place, perhaps for the time being remove the @deprecated. Since there is not yet always a clear alternative.

pveentjer commented:
I have removed the deprecated tags since there currently is not yet a full replacement for these methods (something of the later phases). I added documentation to the call method.
